sed/io.py

```diff
@@ -1 +1,163 @@
+        except ValueError:
+            pass
+
+    return dictionary
+
+
+DEFAULT_UNITS = {
+    'X': 'step',
+    'Y': 'step',
+    't': 'step',
+    'tofVoltage': 'V',
+    'extractorVoltage': 'V',
+    'extractorCurrent': 'A',
+    'cryoTemperature': 'K',
+    'sampleTemperature': 'K',
+    'dldTimeBinSize': 'ns',
+    'delay': 'ps',
+    'timeStamp': 's',
+    'energy': 'eV',
+    'kx': '1/A',
+    'ky': '1/A'}
+
+def res_to_xarray(res, bin_names, bin_axes, metadata=None):
+    """ creates a BinnedArray (xarray subclass) out of the given np.array
+    Parameters:
+        res: np.array
+            nd array of binned data
+        bin_names (list): list of names of the binned axes
+        bin_axes (list): list of np.arrays with the values of the axes
+    Returns:
+        ba: BinnedArray (xarray)
+            an xarray-like container with binned data, axis, and all available metadata
+    """
+    dims = bin_names
+    coords = {}
+    for name, vals in zip(bin_names, bin_axes):
+        coords[name] = vals
+
+    xres = xr.DataArray(res, dims=dims, coords=coords)
+
+    for name in bin_names:
+        try:
+            xres[name].attrs['unit'] = DEFAULT_UNITS[name]
+        except KeyError:
+            pass
+
+    xres.attrs['units'] = 'counts'
+    xres.attrs['long_name'] = 'photoelectron counts'
+
+    if metadata is not None:
+        xres.attrs['metadata'] = metadata
+
+    return xres
+
+def xarray_to_h5(data, faddr, mode='w'):
+    """ Save xarray formatted data to hdf5
+    Args:
+        data (xarray.DataArray): input data
+        faddr (str): complete file name (including path)
+        mode (str): hdf5 read/write mode
+    Returns:
+    """
+    with h5py.File(faddr, mode) as h5File:
+
+        print(f'saving data to {faddr}')
+
+        # Saving data
+
+        ff = h5File.create_group('binned')
+
+        # make a single dataset
+        ff.create_dataset('BinnedData', data=data.data)
+
+        # Saving axes
+        aa = h5File.create_group("axes")
+        ax_n = 0
+        for binName in data.dims:
+            ds = aa.create_dataset(f'ax{ax_n}', data=data.coords[binName])
+            ds.attrs['name'] = binName
+            ax_n += 1
+
+
+        if ('metadata' in data.attrs and isinstance(data.attrs['metadata'], dict)):
+            meta_group = h5File.create_group('metadata')
+
+            recursive_write_metadata(meta_group, data.attrs['metadata'])
+                   
+    print('Saving complete!')
+
+
+
+def h5_to_xarray(faddr, mode='r'):
+    """ Rear xarray data from formatted hdf5 file
+    Args:
+        faddr (str): complete file name (including path)
+        mode (str): hdf5 read/write mode
+    Returns:
+        xarray (xarray.DataArray): output xarra data
+    """
+    with h5py.File(faddr, mode) as h5_file:
+        # Reading data array
+        try:
+            data = h5_file['binned']['BinnedData']
+        except KeyError:
+            print("Wrong Data Format, data not found")
+            raise
```

Comment:
why to re-raise the exception when you can print the error directly in the print statement?

Reply:
You mean in the raise Exception("") statement? I want to create an error with program abortion, but provide a meaningful error message. I understand that is the way to do it?

Reply:
Yep. You use the `try...except` block but then, when the `try` fails due to a KeyError, you print the message and then "re-raise" the exception.
Why not using :
```
except KeyError as e:
    print(f"Wrong data format. Expected ... but got {e}")
```

Reply:
But the problem with this is, that it consumes the Exception, and continues code execution, no? This condition should be a fatal abort condition, as the data format does not fit for loading the file. So I want to abort with raising a fitting exception. I understood this is how I do that.

Reply:
I agree that you still need the `raise` inside the `except` as you want to break code execution.
I whould however print the error message you want to show inside the error. For example:
```
try:
    data = h5_file['binned']['BinnedData']
except KeyError as e:
    raise KeyError(f"Wrong data format. The HDF5 file selected does not match the expected structure. \n Original error: {e}")
```